File: swift/Sources/FieldUICore/Forces/NaturalForces.swift
import Foundation
import simd

// MARK: - Natural primitives (§20.10)
//
// Real field laws, not the designed UI falloffs. The canonical nine are *designed*: finite
// range, soft (1 − d/d_max)ⁿ falloff. These are *natural*: a true softened inverse-square
// law in the sim unit system. `gravity` and `charge` are the same kernel — only the source
// scalar differs (mass ≥ 0 vs. signed charge).
//
// Opt-in: a body only feels them via the "gravity"/"charge" tokens, so registering them
// changes nothing in a field that doesn't ask for them.

/// Chargeable bodies (Stage C2): accumulated charge Q = b.d (the eased density) amplifies
/// the radiated field — as the element charges up, it radiates up to (1 + Q_GAIN)× its base.
let Q_GAIN: Float = 1.5
let DIPOLE_MIN_SEP: Float = 8    // below this the box gives no usable dipole axis
let DIPOLE_MIN_REACH: Float = 60 // synthesized pole reach floor (covers range-0 / point bodies)

/// The body's dipole field at a world point (the visual/structure field, Stage B): the
/// two-pole superposition scaled by the source magnitude `s`. Shared by `magnetism` (the
/// bar magnet, rendered but not followed) and `charge`'s dipole rendering.
func bodyDipole(_ b: Body, at point: Vec3, s: Float) -> Vec3 {
    var (a, c) = polePair(AxisBox(box: b.box, heading: b.heading, spin: b.spin))
    let sep = simd_length(a.position - c.position)
    // synthesize when the box gives no usable separation — every magnetism/charge source
    // then reads as a dipole regardless of size. Pixel floors cover range-0/global bodies.
    if sep < max(b.range * 0.06, DIPOLE_MIN_SEP) {
        let half = max(b.range * 0.18, DIPOLE_MIN_REACH)
        let sgn: Float = b.spin < 0 ? -1 : 1
        a = Pole(position: b.center + b.heading * half, charge: sgn)
        c = Pole(position: b.center - b.heading * half, charge: -sgn)
    }
    let sq = s * (1 + Q_GAIN * b.d) // charged elements radiate a stronger field
    return dipoleField(poles: [a, c], at: point) * sq
}

/// The radial monopole field of a single point charge (the electric field, §20.3):
/// straight field lines out of a positive source, into a negative one.
func bodyMonopole(_ b: Body, at point: Vec3, s: Float) -> Vec3 {
    let d3 = point - b.center
    let d = max(simd_length(d3), EPS)
    let sgn: Float = b.spin < 0 ? -1 : 1
    let mag = (sgn * s * (1 + Q_GAIN * b.d)) / (d * d) // 1/d², signed by polarity
    return (d3 / d) * mag
}

/// The radial gravitational field at a world point (Stage B) — always toward the mass.
func bodyGravityField(_ b: Body, at point: Vec3) -> Vec3 {
    let d3 = b.center - point // toward the body — gravity attracts
    let d = max(simd_length(d3), EPS)
    let mag = (b.M * (1 + Q_GAIN * b.d)) / (d * d)
    return (d3 / d) * mag
}

/// The shared softened inverse-square kernel (§20.10): `s / (d² + ε²)` along the unit
/// vector toward the body, then clamp speed to the unit system's `c`.
///
/// Plummer softening ε = r_s = 2GM/c² keeps the force finite at the core while staying a
/// true 1/d² law far out. `s` is the signed source strength: +GM for gravity (always
/// attractive), −σ·q·GM for charge (like repels, opposite attracts).
public func inverseSquare(_ b: Body, _ p: Particle, _ e: Env, s: Float) {
    if e.dist >= b.range { return } // practical cutoff radius (an N-body softening too)
    let rs = (2 * e.G * b.M) / (e.c * e.c) // Schwarzschild radius → softening ε
    let f = s / (e.dist * e.dist + rs * rs)
    p.velocity += (e.vector / e.dist) * f
    clampToC(p, e.c)
}

/// Clamp a particle's speed to the unit system's `c` — the hard velocity cap that IS the
/// in-sim speed of light (§20.10). Shared by the natural primitives.
@inline(__always)
func clampToC(_ p: Particle, _ c: Float) {
    let sp = simd_length(p.velocity)
    if sp > c { p.velocity *= c / sp }
}

/// §20.10 — true softened inverse-square: F = GM·d̂/(d²+ε²), always attractive.
public struct GravityForce: Force {
    public let token = "gravity"
    public let label = "Gravity"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        inverseSquare(b, p, e, s: e.G * b.M) // GM, mass-sourced (M ≥ 0 → pulls in)
    }

    /// The inward radial gravitational field (Stage B): renderable structure `fieldflow`
    /// can follow. `apply` is unchanged — a field line is not always a particle path.
    public func field(body b: Body, at point: Vec3) -> Vec3? {
        bodyGravityField(b, at: point)
    }
}

/// §20.3/§20.10 — the signed sibling of gravity; same kernel, sign sets direction.
public struct ChargeForce: Force {
    public let token = "charge"
    public let label = "Charge"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let q = p.charge ?? 0 // neutral matter ignores charge fields
        if q == 0 { return }
        // F = σ·q·GM/(d²+ε²); negated for the inward-pointing kernel so like signs repel.
        inverseSquare(b, p, e, s: -(b.spin * q * e.G * b.M))
    }

    /// The radial electric field (Stage B): straight lines, OUT of a + source, IN to a −.
    /// A lone charge is a monopole (unlike a magnet's dipole).
    public func field(body b: Body, at point: Vec3) -> Vec3? {
        bodyMonopole(b, at: point, s: b.M)
    }
}

/// §20.10 — the Lorentz force on a moving charge: curves a particle's path **without doing
/// work** — speed is preserved, only the heading turns. `spin` sets the field's sense,
/// `strength` is |B|, graded by a (1 − d/r) falloff so the curl eases to zero at the rim.
///
/// Implemented as an exact rotation about the field axis by θ = q·spin·B·falloff per frame
/// — preserves |v| to floating-point precision (the Euler form accumulates speed). In 3D
/// the axis is the plane normal; a volumetric B along the body's heading is a one-line swap.
public struct MagnetismForce: Force {
    public let token = "magnetism"
    public let label = "Magnetism"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if e.dist >= b.range { return } // inside the field region
        let q = p.charge ?? 0
        if q == 0 { return } // the Lorentz force needs charge
        let falloff = 1 - e.dist / b.range // ∈ (0, 1] inside the region
        let theta = q * b.spin * b.strength * falloff
        p.velocity = simd_quatf(angle: theta, axis: PLANE_AXIS).act(p.velocity)
    }

    /// The dipole structure of B (Stage B). Rendered as field lines; particles curve
    /// perpendicular to it rather than following it.
    public func field(body b: Body, at point: Vec3) -> Vec3? {
        bodyDipole(b, at: point, s: b.strength)
    }
}

/// The Langevin noise amplitude σ = √(2·k_B·T·γ) (§20.10). In sim units k_B = γ = 1, so
/// σ = √(2T); negative T is floored to 0 (no imaginary kicks).
public func thermalSigma(_ T: Float) -> Float {
    sqrt(2 * max(0, T))
}

/// §20.10 — `thermal`: Langevin/Brownian agitation, the *honest* wander. Each frame a
/// charge-free Gaussian kick v += σ·ξ jiggles matter, with σ = √(2T). Paired with drag
/// it's a thermostat — fluctuation–dissipation, the swarm equilibrates at temperature T.
/// Box–Muller turns two uniforms into one isotropic kick (gaining a z component in a
/// volumetric field).
public struct ThermalForce: Force {
    public let token = "thermal"
    public let label = "Thermal"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if e.dist >= b.range { return }
        let falloff = 1 - e.dist / b.range // localized: hotter nearer the source
        let sigma = thermalSigma(b.strength * falloff)
        if sigma == 0 { return }
        // Box–Muller: (u1, u2) → one isotropic N(0,1) pair, scaled by σ.
        let u1 = max(Float.random(in: 0..<1), 1e-9) // avoid log(0)
        let mag = sigma * sqrt(-2 * log(u1))
        let ang = 2 * Float.pi * Float.random(in: 0..<1)
        p.velocity.x += mag * cos(ang)
        p.velocity.y += mag * sin(ang)
        if e.volume.z > 0 {
            let u2 = max(Float.random(in: 0..<1), 1e-9)
            p.velocity.z += sigma * sqrt(-2 * log(u2)) * cos(2 * Float.pi * Float.random(in: 0..<1))
        }
        if b.isEngaged { p.heat = max(p.heat, falloff * 0.4) }
        clampToC(p, e.c)
    }
}

/// §20.10 — `collide`: elastic pairwise collision (granular / billiard). For each
/// neighbour whose sphere overlaps and that is *approaching*, the pair exchanges normal
/// momentum symmetrically in one pass — momentum-conserving and order-independent.
/// `strength` is the restitution e ∈ [0,1]. Class [B] — uses env.neighbors.
public struct CollideForce: Force {
    public let token = "collide"
    public let label = "Collide"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if e.dist >= b.range { return } // collisions resolve within the body's region
        let restitution = clamp(b.strength, 0, 1)
        let pr = max(1, p.size)
        for q in e.neighbors(p, pr * 4) {
            let qr = max(1, q.size)
            let n = p.position - q.position
            let d = simd_length(n)
            if d >= pr + qr || d < 1e-6 { continue } // not in contact
            let u = n / d
            let relN = simd_dot(p.velocity - q.velocity, u)
            if relN >= 0 { continue } // separating already → no impulse
            // equal-and-opposite half-impulses so the pass is symmetric; after this the
            // pair is separating, so q's own apply skips it.
            let j = (1 + restitution) * 0.5 * relN
            p.velocity -= u * j
            q.velocity += u * j
        }
    }
}

/// §20.10 — `diffuse` (class [C], over the scalar `grid`): the pheromone/stigmergy
/// field. Each frame a particle lays a mark into the shared `diffuse` grid (which the
/// engine blurs via ∂φ/∂t = D∇²φ) and steers *up* the local gradient, following the
/// smeared trail toward where matter has gathered. `strength` sets both the deposit and
/// the follow gain. Self-organizing trails emerge from the deposit↔blur↔follow loop.
public struct DiffuseForce: Force {
    public let token = "diffuse"
    public let label = "Diffuse"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if e.dist >= b.range { return }
        let g = e.grid("diffuse")
        g.deposit(at: p.position, amount: b.strength) // lay a mark
        let grad = g.gradient(at: p.position) // follow the blurred trail up-gradient
        p.velocity += grad * b.strength
    }
}

/// Frames between emitted shocks while a propagate body is engaged. A *pulse train* — not
/// a continuous drip — is what keeps it a travelling wave: between pulses the grid
/// radiates and damps, so no standing bump builds at the source.
let WAVE_PULSE_PERIOD = 12
/// How hard a passing wavefront carries matter outward (radiation pressure gain).
let WAVE_PUSH: Float = 7

/// §20.10 — `propagate` (class [C], over a wave-mode `grid`): a travelling disturbance,
/// ∂²φ/∂t² = c²∇²φ. An engaged body injects an impulsive shock at its centre (via the
/// body-level `source` hook, once per frame), and the grid carries it outward as a real
/// expanding ring. Matter **rides the front out** — radiation pressure, not an inward pull.
public struct PropagateForce: Force {
    public let token = "propagate"
    public let label = "Propagate"
    public init() {}

    public func source(body b: Body, env e: Env) {
        if !b.isEngaged { return } // only an engaged body emits
        if e.frameN % WAVE_PULSE_PERIOD != 0 { return } // a shock train, once per period
        e.grid("wave-propagate").deposit(at: b.center, amount: b.strength) // 'wave…' → wave stepping
    }

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if e.dist >= b.range { return }
        let g = e.grid("wave-propagate")
        let grad = g.gradient(at: p.position)
        let act = simd_length(grad) // wavefront activity — steep where a front is passing
        if act < 1e-6 { return } // no front here → coast
        // ride the front: pushed radially OUTWARD (env.vector points toward the body, so negate).
        let u = -e.vector / e.dist
        p.velocity += u * (act * b.strength * WAVE_PUSH)
        clampToC(p, e.c)
    }
}

/// Memory (class [C], over a slow-decaying `memory` grid) — the field remembers.
/// Each frame a particle lays occupancy where it sits, into a grid that barely blurs and
/// fades slowly; the body's pull is then amplified by how worn the spot is: M(x) += λ and
/// the effective force ×= (1 + μ·M) — frequently-travelled routes deepen and pull harder.
public struct MemoryForce: Force {
    public let token = "memory"
    public let label = "Memory"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if e.dist >= b.range { return }
        let g = e.grid("memory") // 'memory' name → slow-decay stepping
        g.deposit(at: p.position, amount: b.strength * 0.15) // wear the path where matter sits
        let amp = 1 + 0.5 * g.sample(at: p.position) // worn paths pull harder (1 + μ·M)
        let f = pow(1 - e.dist / b.range, 2) * b.strength * 0.5 * amp
        p.velocity += (e.vector / e.dist) * f
    }
}

/// The natural primitives, in spec order (§20.10).
public func naturalForces() -> [any Force] {
    [GravityForce(), ChargeForce(), MagnetismForce(), ThermalForce(), CollideForce(),
     DiffuseForce(), PropagateForce(), MemoryForce()]
}

File: android/fundamental-core/src/main/kotlin/com/fundamental/core/forces/NaturalForces.kt
package com.fundamental.core.forces

import com.fundamental.core.engine.AxisBox
import com.fundamental.core.engine.Body
import com.fundamental.core.engine.Env
import com.fundamental.core.engine.Force
import com.fundamental.core.engine.Particle
import com.fundamental.core.engine.Pole
import com.fundamental.core.engine.dipoleField
import com.fundamental.core.engine.polePair
import com.fundamental.core.math.EPS
import com.fundamental.core.math.Vec3
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.random.Random

// Natural primitives (§20.10) — the Kotlin port of swift/Sources/FundamentalCore/Forces/NaturalForces.swift.
// Real field laws (true softened inverse-square), not the designed UI falloffs. Opt-in via tokens.
//
// gravity/charge/magnetism are deterministic and exactly testable; thermal (RNG), collide (neighbors),
// and the grid-backed three (diffuse/propagate/memory) are verified behaviorally against a constructed Env.
//
// The visual `field()` structure hooks (the dipole/monopole renderable fields) are implemented for the
// monopole sources (gravity, charge); magnetism's dipole field defers to the geometry/integrator port.

/** Charged elements radiate up to (1 + Q_GAIN)× their base field as they charge up (Stage C2). */
private const val Q_GAIN: Float = 1.5f
private const val DIPOLE_MIN_SEP: Float = 8f // below this the box gives no usable dipole axis
private const val DIPOLE_MIN_REACH: Float = 60f // synthesized pole reach floor (point bodies)

private val TWO_PI: Float = (2.0 * PI).toFloat()

/** Clamp a particle's speed to the unit system's `c` — the in-sim speed of light (§20.10). */
fun clampToC(p: Particle, c: Float) {
    val sp = p.velocity.length()
    if (sp > c) p.velocity *= c / sp
}

/**
 * The shared softened inverse-square kernel (§20.10): `s / (d² + ε²)` along the unit vector toward the
 * body, then clamp speed to `c`. Plummer softening ε = r_s = 2GM/c² keeps it finite at the core.
 */
fun inverseSquare(b: Body, p: Particle, e: Env, s: Float) {
    if (e.dist >= b.range) return
    val rs = (2f * e.G * b.M) / (e.c * e.c) // Schwarzschild radius → softening ε
    val f = s / (e.dist * e.dist + rs * rs)
    p.velocity += (e.vector / e.dist) * f
    clampToC(p, e.c)
}

/** The Langevin noise amplitude σ = √(2T); negative T floored to 0. */
fun thermalSigma(t: Float): Float = sqrt(2f * max(0f, t))

/** The inward radial gravitational field at a world point (Stage B) — always toward the mass. */
private fun bodyGravityField(b: Body, point: Vec3): Vec3 {
    val d3 = b.center - point // toward the body — gravity attracts
    val d = max(d3.length(), EPS)
    val mag = (b.M * (1f + Q_GAIN * b.d)) / (d * d)
    return (d3 / d) * mag
}

/**
 * The body's dipole field at a world point (Stage B): the two-pole superposition scaled by `s`. Shared
 * by magnetism (the bar magnet) and charge's dipole rendering. Synthesizes poles when the box gives no
 * usable separation, so every source reads as a dipole regardless of size.
 */
private fun bodyDipole(b: Body, point: Vec3, s: Float): Vec3 {
    var (a, c) = polePair(AxisBox(b.box, b.heading, b.spin))
    val sep = (a.position - c.position).length()
    if (sep < max(b.range * 0.06f, DIPOLE_MIN_SEP)) {
        val half = max(b.range * 0.18f, DIPOLE_MIN_REACH)
        val sgn = if (b.spin < 0f) -1f else 1f
        a = Pole(b.center + b.heading * half, sgn)
        c = Pole(b.center - b.heading * half, -sgn)
    }
    val sq = s * (1f + Q_GAIN * b.d) // charged elements radiate a stronger field
    return dipoleField(listOf(a, c), point) * sq
}

/** The radial monopole field of a single point charge (§20.3): out of +, into −. */
private fun bodyMonopole(b: Body, point: Vec3, s: Float): Vec3 {
    val d3 = point - b.center
    val d = max(d3.length(), EPS)
    val sgn = if (b.spin < 0f) -1f else 1f
    val mag = (sgn * s * (1f + Q_GAIN * b.d)) / (d * d)
    return (d3 / d) * mag
}

/** §20.10 — true softened inverse-square: F = GM·d̂/(d²+ε²), always attractive. */
class GravityForce : Force {
    override val token = "gravity"
    override val label = "Gravity"
    override fun apply(body: Body, particle: Particle, env: Env) =
        inverseSquare(body, particle, env, env.G * body.M) // GM, mass-sourced

    override fun field(body: Body, at: Vec3): Vec3 = bodyGravityField(body, at)
}

/** §20.3/§20.10 — the signed sibling of gravity; same kernel, sign sets direction. */
class ChargeForce : Force {
    override val token = "charge"
    override val label = "Charge"
    override fun apply(body: Body, particle: Particle, env: Env) {
        val q = particle.charge ?: 0f // neutral matter ignores charge fields
        if (q == 0f) return
        // F = σ·q·GM/(d²+ε²); negated for the inward kernel so like signs repel.
        inverseSquare(body, particle, env, -(body.spin * q * env.G * body.M))
    }

    override fun field(body: Body, at: Vec3): Vec3 = bodyMonopole(body, at, body.M)
}

/**
 * §20.10 — the Lorentz force on a moving charge: curves the path without doing work (speed preserved,
 * only the heading turns). Implemented as an exact rotation about the plane axis by θ = q·spin·B·falloff.
 */
class MagnetismForce : Force {
    override val token = "magnetism"
    override val label = "Magnetism"
    override fun apply(body: Body, particle: Particle, env: Env) {
        if (env.dist >= body.range) return
        val q = particle.charge ?: 0f
        if (q == 0f) return // the Lorentz force needs charge
        val falloff = 1f - env.dist / body.range // ∈ (0, 1] inside the region
        val theta = q * body.spin * body.strength * falloff
        particle.velocity = particle.velocity.rotatedAboutZ(theta)
    }

    /** The dipole structure of B (Stage B). Particles curve perpendicular to it, not along it. */
    override fun field(body: Body, at: Vec3): Vec3 = bodyDipole(body, at, body.strength)
}

/**
 * §20.10 — `thermal`: Langevin/Brownian agitation. Each frame a charge-free Gaussian kick v += σ·ξ
 * jiggles matter, σ = √(2T). Box–Muller turns two uniforms into one isotropic kick.
 */
class ThermalForce : Force {
    override val token = "thermal"
    override val label = "Thermal"
    override fun apply(body: Body, particle: Particle, env: Env) {
        if (env.dist >= body.range) return
        val falloff = 1f - env.dist / body.range // hotter nearer the source
        val sigma = thermalSigma(body.strength * falloff)
        if (sigma == 0f) return
        val u1 = max(Random.nextFloat(), 1e-9f) // avoid log(0)
        val mag = sigma * sqrt(-2f * ln(u1))
        val ang = TWO_PI * Random.nextFloat()
        var v = particle.velocity
        v = Vec3(v.x + mag * cos(ang), v.y + mag * sin(ang), v.z)
        if (env.volume.z > 0f) {
            val u2 = max(Random.nextFloat(), 1e-9f)
            v = Vec3(v.x, v.y, v.z + sigma * sqrt(-2f * ln(u2)) * cos(TWO_PI * Random.nextFloat()))
        }
        particle.velocity = v
        if (body.isEngaged) particle.heat = max(particle.heat, falloff * 0.4f)
        clampToC(particle, env.c)
    }
}

/**
 * §20.10 — `collide`: elastic pairwise collision. For each approaching, overlapping neighbour the pair
 * exchanges normal momentum symmetrically in one pass — momentum-conserving and order-independent.
 */
class CollideForce : Force {
    override val token = "collide"
    override val label = "Collide"
    override fun apply(body: Body, particle: Particle, env: Env) {
        if (env.dist >= body.range) return
        val restitution = com.fundamental.core.math.clamp(body.strength, 0f, 1f)
        val pr = max(1f, particle.size)
        for (q in env.neighbors(particle, pr * 4f)) {
            val qr = max(1f, q.size)
            val n = particle.position - q.position
            val d = n.length()
            if (d >= pr + qr || d < 1e-6f) continue // not in contact
            val u = n / d
            val relN = (particle.velocity - q.velocity).dot(u)
            if (relN >= 0f) continue // separating already → no impulse
            val j = (1f + restitution) * 0.5f * relN
            particle.velocity -= u * j
            q.velocity += u * j
        }
    }
}

/**
 * §20.10 — `diffuse` (over the scalar `diffuse` grid): the pheromone/stigmergy field. Lay a mark, then
 * steer up the blurred gradient toward where matter has gathered. `strength` sets deposit + follow gain.
 */
class DiffuseForce : Force {
    override val token = "diffuse"
    override val label = "Diffuse"
    override fun apply(body: Body, particle: Particle, env: Env) {
        if (env.dist >= body.range) return
        val g = env.grid("diffuse")
        g.deposit(particle.position, body.strength) // lay a mark
        val grad = g.gradient(particle.position) // follow the blurred trail up-gradient
        particle.velocity += grad * body.strength
    }
}

/** Frames between emitted shocks while a propagate body is engaged. */
private const val WAVE_PULSE_PERIOD = 12

/** How hard a passing wavefront carries matter outward (radiation pressure gain). */
private const val WAVE_PUSH: Float = 7f

/**
 * §20.10 — `propagate` (over a wave-mode grid): a travelling disturbance, ∂²φ/∂t² = c²∇²φ. An engaged
 * body injects an impulsive shock at its centre once per period; matter rides the front out.
 */
class PropagateForce : Force {
    override val token = "propagate"
    override val label = "Propagate"

    override fun source(body: Body, env: Env) {
        if (!body.isEngaged) return // only an engaged body emits
        if (env.frameN % WAVE_PULSE_PERIOD != 0) return // a shock train, once per period
        env.grid("wave-propagate").deposit(body.center, body.strength) // 'wave…' → wave stepping
    }

    override fun apply(body: Body, particle: Particle, env: Env) {
        if (env.dist >= body.range) return
        val g = env.grid("wave-propagate")
        val grad = g.gradient(particle.position)
        val act = grad.length() // wavefront activity — steep where a front is passing
        if (act < 1e-6f) return // no front here → coast
        val u = (env.vector / env.dist) * -1f // ride the front: radially OUTWARD
        particle.velocity += u * (act * body.strength * WAVE_PUSH)
        clampToC(particle, env.c)
    }
}

/**
 * §20.10 — `memory` (over a slow-decaying grid): the field remembers. Lay occupancy where matter sits;
 * the body's pull is amplified by how worn the spot is — frequently-travelled routes pull harder.
 */
class MemoryForce : Force {
    override val token = "memory"
    override val label = "Memory"
    override fun apply(body: Body, particle: Particle, env: Env) {
        if (env.dist >= body.range) return
        val g = env.grid("memory") // 'memory' name → slow-decay stepping
        g.deposit(particle.position, body.strength * 0.15f) // wear the path
        val amp = 1f + 0.5f * g.sample(particle.position) // worn paths pull harder (1 + μ·M)
        val f = (1f - env.dist / body.range).pow(2f) * body.strength * 0.5f * amp
        particle.velocity += (env.vector / env.dist) * f
    }
}

/** The natural primitives, in spec order (§20.10). */
fun naturalForces(): List<Force> = listOf(
    GravityForce(), ChargeForce(), MagnetismForce(), ThermalForce(), CollideForce(),
    DiffuseForce(), PropagateForce(), MemoryForce(),
)
